src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java

```diff
@@ -699,86 +699,67 @@ private void setCreatedWorkspacePermissions(
-				} catch (MongoException me) {
-					throw new WorkspaceCommunicationException(ERR_DB_COMM, me);
-				}
-				if (ur.getModifiedCount() == 1) { //ok, it worked
-					success = true;
-					continue;
-				}
-				//add the key/value pair to the array
-				time = Instant.now();
-				try {
-					ur = wsmongo.getCollection(COL_WORKSPACES).updateOne(
-							new Document(Fields.WS_ID, rwsi.getID())
-									.append(mkey, new Document("$nin", Arrays.asList(key))),
-							new Document(
-									"$push", new Document(Fields.WS_META,
-											new Document(Fields.META_KEY, key)
-													.append(Fields.META_VALUE, value)))
-									.append("$set", new Document(
-											Fields.WS_MODDATE, Date.from(time))));
-
-				} catch (MongoException me) {
-					throw new WorkspaceCommunicationException(ERR_DB_COMM, me);
-				}
-				if (ur.getModifiedCount() == 1) { //ok, it worked
-					success = true;
-				}
-				/* amazingly, someone added that key to the metadata between the
-				   two calls above, so here we go again on our own
-				   Should be impossible to get stuck in a loop, but if so add
-				   counter and throw error if > 3 or something
-				 */
+		while (time == null) {
+			final Map<String, Object> ws = query.queryWorkspace(rwsi, FLDS_WS_META);
+			@SuppressWarnings("unchecked")
+			final List<Map<String, String>> mlist = (List<Map<String, String>>)
+					ws.get(Fields.WS_META);
+			final Map<String, String> updatedMeta = metaMongoArrayToHash(mlist);
+			updatedMeta.putAll(newMeta.getMetadata());
+			try {
+				WorkspaceUserMetadata.checkMetadataSize(updatedMeta);
+			} catch (MetadataException me) {
+				throw new IllegalArgumentException(String.format(
+						"Updated metadata exceeds allowed size of %sB",
+						WorkspaceUserMetadata.MAX_METADATA_SIZE));
 			}
+			final Document query = new Document(Fields.WS_ID, rwsi.getID())
+					.append(Fields.WS_META, mlist);
+			final Document metaUpdate = new Document(
+					Fields.WS_META, metaHashToMongoArray(updatedMeta));
+			time = _internal_setWorkspaceMeta(attempts, 5, query, metaUpdate);
+			attempts++;
 		}
 		return time;
 	}
+	
+	// split the method for testing purposes
+	private Instant _internal_setWorkspaceMeta(
+			final int attempts,
+			final int maxattempts,
+			final Document query,
+			final Document metaUpdate)
+			throws WorkspaceCommunicationException {
+		try {
+			final Instant time = clock.instant();
+			// only match if the metadata we pulled from the db is stil the same, so we don't
+			// clobber any interleaving changes
+			final UpdateResult ur = wsmongo.getCollection(COL_WORKSPACES).updateOne(
+					query,
+					new Document("$set", metaUpdate.append(Fields.WS_MODDATE, Date.from(time))));
+			if (ur.getModifiedCount() == 1) { //ok, it worked
+				return time;
+			} else if (attempts >= maxattempts) {
+				throw new WorkspaceCommunicationException(
+						String.format("Failed to update metadata %s times", attempts));
+			} else {
+				return null;
+			}
```

Comment:
if you were having an anti-indentation day, you could do:

```
			if (ur.getModifiedCount() == 1) { //ok, it worked
 				return time;
 			}
 			if (attempts >= maxattempts) {
 				throw new WorkspaceCommunicationException(
 						String.format("Failed to update metadata %s times", attempts));
 			}
                        return null;
```

Reply:
I actually prefer the `else`s here, makes things a bit easier to read for me. Not a big deal though